### PR TITLE
MET-4159 Make Http and Zip file harvesting Asynchronous

### PR DESCRIPTION
**Pull request by JoanaCMS**

### CODE REVIEW

File: src/test/java/eu/europeana/metis/sandbox/service/workflow/HarvestServiceImplTest.java [hunk 1]
```diff
@@ -1,239 +1,336 @@
...
+import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
 import org.mockito.Mock;
-import org.springframework.mock.web.MockMultipartFile;
 import org.springframework.test.context.junit.jupiter.SpringExtension;
-import org.springframework.util.CollectionUtils;
-import org.zeroturnaround.zip.ZipException;
```
Comment (jeortizquan) [inline src/test/java/eu/europeana/metis/sandbox/service/workflow/HarvestServiceImplTest.java:42]: Remove unused imports.

File: src/test/java/eu/europeana/metis/sandbox/service/workflow/HarvestServiceImplTest.java [hunk 2]
```diff
@@ -1,239 +1,336 @@
...
         when(recordRepository.save(any(RecordEntity.class))).thenReturn(recordEntity);
 
-        var harvestContent = harvestService.harvestOaiRecordHeader("1", oaiHarvestData, record);
-
-        assertEquals("1", harvestContent.getRecord().getDatasetId());
-        assertEquals(Country.NETHERLANDS, harvestContent.getRecord().getCountry());
-        assertEquals(Language.NL, harvestContent.getRecord().getLanguage());
-        assertTrue(CollectionUtils.isEmpty(harvestContent.getErrors()));
-        assertEquals("record", new String(harvestContent.getRecord().getContent(), StandardCharsets.UTF_8));
+        harvestService.harvestOaiPmh("datasetId", createMockEncapsulatedRecord(), oaiHarvestData);
```
Comment (jeortizquan) [inline src/test/java/eu/europeana/metis/sandbox/service/workflow/HarvestServiceImplTest.java:287]: remove duplicate code assertions can be extracted to a method. ```assertHarvestProcess```

```
        verify(asyncRecordPublishService, times(XXX)).publishWithoutXslt(captorRecordInfo.capture(), eq(Step.XXX));
        verify(datasetService).updateNumberOfTotalRecord(eq("datasetId"), eq(XXX));
        assertTrue(captorRecordInfo.getAllValues().stream().allMatch(x -> x.getRecord().getDatasetId().equals("datasetId")));
        assertTrue(captorRecordInfo.getAllValues().stream().allMatch(x -> x.getRecord().getContent() != null));
        assertEquals(Country.NETHERLANDS, captorRecordInfo.getValue().getRecord().getCountry());
        assertEquals(Language.NL, captorRecordInfo.getValue().getRecord().getLanguage());
        assertEquals("datasetName", captorRecordInfo.getValue().getRecord().getDatasetName());
```

**Review by jeortizquan — CHANGES_REQUESTED** [review]
There is some code changes that I think can be made.

File: src/main/java/eu/europeana/metis/sandbox/controller/DatasetController.java [hunk 3]
```diff
@@ -113,17 +106,11 @@ public DatasetIdDto harvestDatasetFromFile(
     checkArgument(namePattern.matcher(datasetName).matches(),
         "dataset name can only include letters, numbers, _ or - characters");
 
-    HarvestContent harvestedRecords =
-        harvestService.harvestZipMultipartFile(dataset);
-
     InputStream xsltFileString = createXsltAsInputStreamIfPresent(xsltFile);
```
[the remark this answers is not in the document]
Comment (stzanakis) [reply src/main/java/eu/europeana/metis/sandbox/controller/DatasetController.java:109]: Done

File: src/main/java/eu/europeana/metis/sandbox/service/dataset/AsyncRecordPublishService.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
+package eu.europeana.metis.sandbox.service.dataset;
+
+import eu.europeana.metis.sandbox.common.Step;
+import eu.europeana.metis.sandbox.domain.RecordInfo;
+
+import java.util.concurrent.CompletableFuture;
+
+public interface AsyncRecordPublishService {
```
[the remark this answers is not in the document]
Comment (stzanakis) [reply src/main/java/eu/europeana/metis/sandbox/service/dataset/AsyncRecordPublishService.java:8]: I agree! Updated.

File: src/main/java/eu/europeana/metis/sandbox/service/dataset/AsyncRecordPublishService.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+   * Async publish to message broker for further processing. This will send messages to 'created`
+   * queue to skip the transformation to edm external step
+   *
+   * @param recordToPublish to publish. Must not be null
+   * @param step The step that was performed
+   * @return {@link CompletableFuture} of the process
+   * @throws NullPointerException     if dataset object is null
+   * @throws IllegalArgumentException if dataset records are empty
+   */
+  CompletableFuture<Void> publishWithoutXslt(RecordInfo recordToPublish, Step step);
```
[the remark this answers is not in the document]
Comment (stzanakis) [reply src/main/java/eu/europeana/metis/sandbox/service/dataset/AsyncRecordPublishService.java:20]: done

File: src/main/java/eu/europeana/metis/sandbox/service/dataset/AsyncRecordPublishService.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+   * Async publish to message broker for further processing. This will send messages to
+   * 'transformationToEdmExternal' queue to go through the transformation to edm external step
+   *
+   * @param recordToPublish to publish. Must not be null
+   * @param step The step that was performed
+   * @return {@link CompletableFuture} of the process
+   * @throws NullPointerException     if dataset object is null
+   * @throws IllegalArgumentException if dataset records are empty
+   */
+  CompletableFuture<Void> publishWithXslt(RecordInfo recordToPublish, Step step);
```
[the remark this answers is not in the document]
Comment (stzanakis) [reply src/main/java/eu/europeana/metis/sandbox/service/dataset/AsyncRecordPublishService.java:32]: done

File: src/main/java/eu/europeana/metis/sandbox/service/dataset/AsyncRecordPublishService.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (stzanakis) [reply src/main/java/eu/europeana/metis/sandbox/service/dataset/AsyncRecordPublishService.java:8]: Come to think of it, this is not required to be asynchronous at all

File: src/main/java/eu/europeana/metis/sandbox/service/workflow/HarvestService.java [hunk 7]
```diff
@@ -1,46 +1,30 @@
...
-   * @param url URL for zip file containing one or more records
-   * @return A HarvestContent object containing the content of the harvest and a bollean indicating
-   * if it reached the max number of records
-   * @throws ServiceException if error processing URL, if URL timeout, if records are empty
+   * @param datasetId The id of the dataset to be harvested
+   * @param recordDataEncapsulated The encapsulation of data to be used to harvest each record
+   * @param oaiHarvestData The object that encapsulate the necessary data for harvesting
    */
-  HarvestContent harvestZipUrl(String url) throws ServiceException;
+  void harvestOaiPmh(String datasetId, Record.RecordBuilder recordDataEncapsulated, OaiHarvestData oaiHarvestData);
```
[the remark this answers is not in the document]
Comment (stzanakis) [reply src/main/java/eu/europeana/metis/sandbox/service/workflow/HarvestService.java:19]: Most of the namings have it OaiPmh in fact.

[the remark this answers is not in the document]
Comment (jeortizquan) [reply src/main/java/eu/europeana/metis/sandbox/service/workflow/HarvestService.java:19]: well it is just a matter of taste :stuck_out_tongue_winking_eye: i think.

File: src/main/java/eu/europeana/metis/sandbox/service/workflow/AsyncHarvestPublishServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+            Record.RecordBuilder recordDataEncapsulated = Record.builder().datasetId(datasetId).datasetName(datasetName).country(country).language(language);
+            return runHarvestAsync(file.getInputStream(), datasetId, recordDataEncapsulated);
+        } catch (IOException e) {
+            throw new ServiceException("Error harvesting records from file " + file.getName(), e);
+        }
+    }
+
+    @Override
+    public CompletableFuture<Void> runHttpHarvestAsync(String url, String datasetName, String datasetId, Country country, Language language){
+        try {
```
[the remark this answers is not in the document]
Comment (stzanakis) [reply src/main/java/eu/europeana/metis/sandbox/service/workflow/AsyncHarvestPublishServiceImpl.java:43]: Absolutely.

File: src/test/java/eu/europeana/metis/sandbox/controller/DatasetControllerTest.java [hunk 9]
```diff
@@ -65,569 +62,358 @@
...
+                        .param("language", IT.name())
+                        .param("url", url))
+                .andExpect(status().isAccepted())
+                .andExpect(jsonPath("$.dataset-id", is("12345")));
+    }
+
+    @Test
+    void processDatasetFromOAI_expectSuccess() throws Exception {
+
+        String url = new URI("http://panic.image.ntua.gr:9000/efg/oai").toString();
```
[the remark this answers is not in the document]
Comment (stzanakis) [reply src/test/java/eu/europeana/metis/sandbox/controller/DatasetControllerTest.java:163]: done

File: src/test/java/eu/europeana/metis/sandbox/controller/DatasetControllerTest.java [hunk 10]
```diff
@@ -65,569 +62,358 @@
...
+                        .param("language", IT.name()))
+                .andExpect(status().isBadRequest())
+                .andExpect(jsonPath("$.message",
+                        is("dataset name can only include letters, numbers, _ or - characters")));
+    }
+
+    @Test
+    void processDatasetFromURL_invalidName_expectFail() throws Exception {
+
+        String url = "zip" + File.separator + "dataset-valid.zip";
```
[the remark this answers is not in the document]
Comment (stzanakis) [reply src/test/java/eu/europeana/metis/sandbox/controller/DatasetControllerTest.java:219]: done

File: src/main/java/eu/europeana/metis/sandbox/service/workflow/HarvestServiceImpl.java [hunk 11]
```diff
@@ -36,54 +41,70 @@
...
+
+      recordHeaderIterator.forEach(recordHeader -> {
+        OaiHarvestData completeOaiHarvestData = new OaiHarvestData(oaiHarvestData.getUrl(),
+                oaiHarvestData.getSetspec(),
+                oaiHarvestData.getMetadataformat(),
+                recordHeader.getOaiIdentifier());
+        currentNumberOfIterations.getAndIncrement();
+
+        if (currentNumberOfIterations.get() > maxRecords) {
+          datasetService.updateRecordsLimitExceededToTrue(datasetId);
```
[the remark this answers is not in the document]
Comment (stzanakis) [reply src/main/java/eu/europeana/metis/sandbox/service/workflow/HarvestServiceImpl.java:82]: Agreed

File: src/test/java/eu/europeana/metis/sandbox/service/workflow/HarvestServiceImplTest.java (hunk 2 above)
Comment (stzanakis) [reply src/test/java/eu/europeana/metis/sandbox/service/workflow/HarvestServiceImplTest.java:287]: done

File: src/test/java/eu/europeana/metis/sandbox/service/dataset/DatasetServiceImplTest.java [hunk 12]
```diff
@@ -46,149 +34,6 @@
...
-        .thenThrow(new RuntimeException("Failed"));
-    when(generatorService.generate(any(DatasetMetadata.class), eq(records)))
-        .thenReturn(dataset);
-
-    assertThrows(ServiceException.class,
-        () -> service.createDataset("name", Country.AUSTRIA, Language.BE, records, false));
-
-    verify(datasetRepository, times(2)).save(any(DatasetEntity.class));
-    verify(publishService, never()).publishWithoutXslt(dataset);
-  }
```
[the remark this answers is not in the document]
Comment (stzanakis) [reply src/test/java/eu/europeana/metis/sandbox/service/dataset/DatasetServiceImplTest.java:191]: Can't find this, probably from another commit.

File: src/test/java/eu/europeana/metis/sandbox/common/TestUtils.java [hunk 13]
```diff
@@ -77,4 +87,35 @@ private boolean safeCheck(ZipEntry zipEntry) {
...
+    public TestHeaderIterator(List<OaiRecordHeader> source) {
+      this.source = source;
+    }
+    @Override
+    public void forEachFiltered(final ReportingIteration<OaiRecordHeader> action,
+                                final Predicate<OaiRecordHeader> filter) {
+      this.source.forEach(action::process);
+    }
+    @Override
+    public void close() {
```
[the remark this answers is not in the document]
Comment (stzanakis) [reply src/test/java/eu/europeana/metis/sandbox/common/TestUtils.java:102]: Why throw? Wouldn't that break the tests?
Same for below.

File: src/main/java/eu/europeana/metis/sandbox/service/workflow/HarvestServiceImpl.java [hunk 14]
```diff
@@ -110,31 +131,79 @@ public RecordInfo harvestOaiRecordHeader(String datasetId, OaiHarvestData oaiHar
...
+      throw new HarvesterException("Could not process path " + exception.get(0).getKey() + ".",
+              exception.get(0).getValue());
     }
-    return new HarvestContent(hasReachedRecordLimit, records);
+
+  }
+
+  private RecordInfo harvestInputStream(InputStream inputStream, String datasetId, Record.RecordBuilder recordToHarvest) throws ServiceException {
+    List<RecordError> recordErrors = new ArrayList<>();
+    RecordEntity recordEntity = recordRepository.save(
```
[the remark this answers is not in the document]
Comment (stzanakis) [reply src/main/java/eu/europeana/metis/sandbox/service/workflow/HarvestServiceImpl.java:185]: It's using the recordEntity.getId for the log in the exception.
In the interest of time, I'm skipping this. Please create a ticket for it and we can have a further look.

File: src/test/java/eu/europeana/metis/sandbox/service/dataset/DatasetServiceImplTest.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (stzanakis) [reply src/test/java/eu/europeana/metis/sandbox/service/dataset/DatasetServiceImplTest.java:191]: Yeah I cannot correlate where this is on the files I have.
If you remember you can add it if it's meaningful.

File: src/test/java/eu/europeana/metis/sandbox/common/TestUtils.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (jeortizquan) [reply src/test/java/eu/europeana/metis/sandbox/common/TestUtils.java:102]: Yes, thanks. I see, my bad, I made a wrong assumption in this part of the code.

**Review by jeortizquan — APPROVED** [review]
Accepted after checking with the team.

**Pull request merged**